File: challenges/contacts/src/main/java/challenges/contacts/Trie.java
package challenges.contacts;

public class Trie {
    private final TrieNode root;

    public Trie() {
        root = new TrieNode();
    }

    public void insert(String word) {
        TrieNode current = root;
        for (char c : word.toCharArray()) {
            if(!current.children.containsKey(c)) {
                current.children.put(c, new TrieNode());
            }
            current.wordCount++;
            current = current.children.get(c);
        }
        current.wordCount++;
        current.endOfWord = true;
    }

    public int startsWith(String word) {
        TrieNode current = root;
        for (char c : word.toCharArray()) {
            current = current.children.get(c);

            if(current == null) {
                return 0;
            }
        }
        return current.wordCount;
    }
}

File: challenges/amazon-demo-1/src/main/kotlin/com/amazon/interview/Trie.kt
package com.amazon.interview

class Trie {
    private val root: TrieNode = TrieNode()

    fun insert(word: String) {
        var current = root
        for (c in word.toCharArray()) {
            if (!current.children.containsKey(c)) {
                current.children[c] = TrieNode()
            }
            current = current.children[c]!!
        }
        if(current.endOfWord) {
            current.count++
        } else {
            current.endOfWord = true
        }
        current.word = word
    }

    fun startsWith(word: String): List<String> {
        val result = mutableListOf<String>()
        var current: TrieNode = root
        for (c in word.toCharArray()) {
            current.children[c]?.let {
                current = it
            } ?: return emptyList()
        }

        recursive(current, result)

        return result
    }

    private fun recursive(node: TrieNode, result : MutableList<String>) {
        if(result.size == 3) {
            return
        }

        if(node.endOfWord) {
            for (i in 0 until node.count) {
                result.add(node.word!!)
                if(result.size == 3) {
                    return
                }
            }
        }

        node.children.forEach {
            recursive(it.value, result)
        }

    }
}
